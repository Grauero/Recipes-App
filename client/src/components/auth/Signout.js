import React from 'react';
import PropTypes from 'prop-types';
import { ApolloConsumer } from 'react-apollo';
import { withRouter } from 'react-router-dom';

function handleSignout(client, history) {
  localStorage.setItem('token', '');
  client.resetStore();
  history.push('/');
}

const Signout = ({ history }) => (
  <ApolloConsumer>
    {client => (
      <button type="button" onClick={() => handleSignout(client, history)}>
        Signout
      </button>
    )}
  </ApolloConsumer>
);

Signout.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

export default withRouter(Signout);
