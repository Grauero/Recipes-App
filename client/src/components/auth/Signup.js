import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { withRouter } from 'react-router-dom';

import Error from '../common/Error';
import { SIGNUP_USER } from '../../queries';

const initialState = {
  username: '',
  email: '',
  password: '',
  passwordConfirmation: ''
};

class Signup extends Component {
  state = { ...initialState };

  clearState = () => this.setState({ ...initialState });

  handleChange = e => this.setState({ [e.target.name]: e.target.value });

  handleSubmit = async (e, signupUser) => {
    e.preventDefault();

    const { data } = await signupUser();
    localStorage.setItem('token', data.signupUser.token);
    await this.props.refetch();

    this.clearState();
    this.props.history.push('/');
  };

  validateForm = () => {
    const {
      username, email, password, passwordConfirmation
    } = this.state;
    const isInvalid = !username
      || !email
      || !password
      || !passwordConfirmation
      || password !== passwordConfirmation;

    return isInvalid;
  };

  render() {
    const {
      username, email, password, passwordConfirmation
    } = this.state;

    return (
      <div className="App">
        <h2 className="App">Signup</h2>
        <Mutation mutation={SIGNUP_USER} variables={{ username, email, password }}>
          {(signupUser, { loading, error }) => (
            <form className="form" onSubmit={e => this.handleSubmit(e, signupUser)}>
              <input
                type="text"
                name="username"
                placeholder="Username"
                value={username}
                onChange={this.handleChange}
              />

              <input
                type="email"
                name="email"
                placeholder="Email Address"
                value={email}
                onChange={this.handleChange}
              />

              <input
                type="password"
                name="password"
                placeholder="Password"
                value={password}
                onChange={this.handleChange}
              />

              <input
                type="password"
                name="passwordConfirmation"
                placeholder="Confirm Password"
                value={passwordConfirmation}
                onChange={this.handleChange}
              />

              <button
                type="submit"
                disabled={loading || this.validateForm()}
                className="button-primary"
              >
                Submit
              </button>

              {error && <Error error={error} />}
            </form>
          )}
        </Mutation>
      </div>
    );
  }
}

Signup.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  refetch: PropTypes.func.isRequired
};

export default withRouter(Signup);
export { Signup };
