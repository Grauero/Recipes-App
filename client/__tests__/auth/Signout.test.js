import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';

import { Signout } from '../../src/components/auth/Signout';

const props = { history: { push: jest.fn() } };
let component;
let wrapper;

beforeEach(() => {
  wrapper = mount(
    <MockedProvider mocks={[]} addTypename={false}>
      <BrowserRouter>
        <Signout {...props} />
      </BrowserRouter>
    </MockedProvider>
  );

  component = wrapper.find(Signout);
});

afterEach(() => {
  wrapper.unmount();
});

it('matches snapshot', () => {
  expect(toJSON(component)).toMatchSnapshot();
});

it('removes token in localStorage when user clicks logout button', () => {
  window.localStorage.setItem('token', 'token');
  component.find('button').simulate('click');

  expect(window.localStorage.getItem('token')).toBeNull();
});

it('redirects to "/" when user clicks logout button', () => {
  component.find('button').simulate('click');

  expect(props.history.push).toHaveBeenCalled();
  expect(props.history.push).toHaveBeenCalledWith('/');
});
