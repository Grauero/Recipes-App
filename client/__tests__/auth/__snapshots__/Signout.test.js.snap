// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders component 1`] = `
<Signout
  history={
    Object {
      "push": [Function],
    }
  }
>
  <ApolloConsumer>
    <button
      onClick={[Function]}
      type="button"
    >
      Signout
    </button>
  </ApolloConsumer>
</Signout>
`;
